diesel/src/pg/connection/mod.rs

```diff
@@ -200,6 +200,32 @@ mod tests {
     }
 
     #[test]
+    fn queries_with_identical_types_and_sql_but_different_bind_types_are_cached_separately() {
+        let connection = connection();
+
+        let query = ::select(AsExpression::<Integer>::as_expression(1)).into_boxed::<Pg>();
+        let query2 = ::select(AsExpression::<VarChar>::as_expression("hi")).into_boxed::<Pg>();
+
+        assert_eq!(Ok(1), query.get_result(&connection));
+        assert_eq!(Ok("hi".to_string()), query2.get_result(&connection));
+        assert_eq!(2, connection.statement_cache.len());
```

Comment:
Just as a sanity check, I'd always put a `assert_eq!(0, connection.statement_cache.len());` (or actually `assert!(connection.statement_cache.is_empty());`) at the top (same below).